src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,903 @@
+                    fmpz_clear(x)
+                    raise TypeError("unsupported coefficient in list")
+                fmpz_mod_poly_set_coeff_fmpz(
+                    poly, i, (<fmpz>(val_fmpz)).val, self.mod.val
+                )
+        fmpz_clear(x)
+        return 0
+
+    cdef set_any_as_fmpz_mod_poly(self, fmpz_mod_poly_t poly, obj):
+        if typecheck(obj, list):
+            return self.set_list_as_fmpz_mod_poly(poly, obj)
+
+        # Convert fmpz_mod to constant poly
+        if typecheck(obj, fmpz_mod):
+            if self.mod != (<fmpz_mod>obj).ctx:
+                raise ValueError("moduli must match")
+            fmpz_mod_poly_set_fmpz(
+                poly, (<fmpz_mod>obj).val, self.mod.val
+            )
+            return 0
+
+        # Reduced fmpz_poly modulo mod
+        if typecheck(obj, fmpz_poly):
+            fmpz_mod_poly_set_fmpz_poly(
+                poly, (<fmpz_poly>obj).val, self.mod.val
+            )
+            return 0
+
+
+        obj = any_as_fmpz(obj)
+        if obj is NotImplemented:
+            return NotImplemented
+        fmpz_mod_poly_set_fmpz(
+            poly, (<fmpz>obj).val, self.mod.val
+        )
+        return 0
+
+    cdef any_as_fmpz_mod_poly(self, obj):
+        # Convert fmpz_mod_poly
+        if typecheck(obj, fmpz_mod_poly):
+            if self != (<fmpz_mod_poly>obj).ctx:
+                raise ValueError("moduli must match")
+            return obj
+        
+        cdef fmpz_mod_poly res
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        check = self.set_any_as_fmpz_mod_poly(res.val, obj)
+        if check is NotImplemented:
+            return NotImplemented
+        
+        res.ctx = self
+        return res
+
+    def __eq__(self, other):
+        # Most often, we expect both `fmpz_mod_poly` to be pointing 
+        # to the same ctx, so this seems the fastest way to check
+        if self is other:
+            return True
+        
+        # If they're not the same object in memory, they may have the
+        # same modulus, which is good enough
+        if typecheck(other, fmpz_mod_poly_ctx):
+            return self.mod == (<fmpz_mod_poly_ctx>other).mod
+        return False
+    
+    def __hash__(self):
+        return hash(self.modulus())
+
+    def __str__(self):
+        return f"Context for fmpz_mod_poly with modulus: {self.modulus()}"
+
+    def __repr__(self):
+        return f"fmpz_mod_poly_ctx({self.modulus()})"
+
+    def __call__(self, val):
+        return fmpz_mod_poly(val, self)
+
+
+cdef class fmpz_mod_poly(flint_poly):
+    """
+    """
+    def __cinit__(self):
+        fmpz_mod_poly_init(self.val, _fmpz_mod_one_ctx)
+
+    def __dealloc__(self):
+        fmpz_mod_poly_clear(self.val, self.ctx.mod.val)
+
+    def __init__(self, val, ctx):
+        if not typecheck(ctx, fmpz_mod_poly_ctx):
+            raise TypeError
+        self.ctx = ctx
+        self.ctx.set_any_as_fmpz_mod_poly(self.val, val)
```

Comment:
I think actually this is still not right because if TypeError is raised then `__dealloc__` is called but `self.ctx` is still null. Also I think that `fmpz_mod_poly_init` needs to be called with the new `ctx`.

This seems to work:
```diff
diff --git a/src/flint/types/fmpz_mod_poly.pyx b/src/flint/types/fmpz_mod_poly.pyx
index 85c4180..b3aee3b 100644
--- a/src/flint/types/fmpz_mod_poly.pyx
+++ b/src/flint/types/fmpz_mod_poly.pyx
@@ -17,12 +17,11 @@ from flint.types.fmpz_poly cimport fmpz_poly
 from flint.flint_base.flint_base cimport flint_poly
 from flint.utils.typecheck cimport typecheck
 
+
 # Global context with modulus one to make `__cinit__` happy within
 # `fmpz_mod_poly`
-cdef fmpz_t _fmpz_one
-cdef fmpz_mod_ctx_t _fmpz_mod_one_ctx
-fmpz_one(_fmpz_one)
-fmpz_mod_ctx_init(_fmpz_mod_one_ctx, _fmpz_one)
+cdef _fmpz_mod_one_poly_ctx = fmpz_mod_poly_ctx(1)
+
 
 cdef class fmpz_mod_poly_ctx:
     """
@@ -264,7 +263,8 @@ cdef class fmpz_mod_poly(flint_poly):
     """
     """
     def __cinit__(self):
-        fmpz_mod_poly_init(self.val, _fmpz_mod_one_ctx)
+        self.ctx = _fmpz_mod_one_poly_ctx
+        fmpz_mod_poly_init(self.val, self.ctx.mod.val)
 
     def __dealloc__(self):
         fmpz_mod_poly_clear(self.val, self.ctx.mod.val)
@@ -272,7 +272,10 @@ cdef class fmpz_mod_poly(flint_poly):
     def __init__(self, val, ctx):
         if not typecheck(ctx, fmpz_mod_poly_ctx):
             raise TypeError
+        # Note sure if clearing is necessary here:
+        fmpz_mod_poly_clear(self.val, self.ctx.mod.val)
         self.ctx = ctx
+        fmpz_mod_poly_init(self.val, self.ctx.mod.val)
         self.ctx.set_any_as_fmpz_mod_poly(self.val, val)
 
     # ---------------- #
```
I assume that using a PyObject like this will mean that it would eventually get freed although I'm not sure if making it cdef prevents that. If it isn't cdef then it possible to replace with a different object from Python which is maybe not a huge problem (it only causes TypeError later rather than segfault).

Reply:
Alternatively it might be simpler just to have a flag to indicate whether initialisation happened. If we were using pointers then we could just use the null value but since the value is inline we need a separate variable. This seems simplest:
```diff
diff --git a/src/flint/types/fmpz_mod_poly.pxd b/src/flint/types/fmpz_mod_poly.pxd
index f2f755e..df6474a 100644
--- a/src/flint/types/fmpz_mod_poly.pxd
+++ b/src/flint/types/fmpz_mod_poly.pxd
@@ -13,5 +13,6 @@ cdef class fmpz_mod_poly_ctx:
 cdef class fmpz_mod_poly(flint_poly):
     cdef fmpz_mod_poly_t val
     cdef fmpz_mod_poly_ctx ctx
+    cdef bint initialized
     cpdef long length(self)
     cpdef long degree(self)
diff --git a/src/flint/types/fmpz_mod_poly.pyx b/src/flint/types/fmpz_mod_poly.pyx
index 85c4180..dec5589 100644
--- a/src/flint/types/fmpz_mod_poly.pyx
+++ b/src/flint/types/fmpz_mod_poly.pyx
@@ -17,12 +17,6 @@ from flint.types.fmpz_poly cimport fmpz_poly
 from flint.flint_base.flint_base cimport flint_poly
 from flint.utils.typecheck cimport typecheck
 
-# Global context with modulus one to make `__cinit__` happy within
-# `fmpz_mod_poly`
-cdef fmpz_t _fmpz_one
-cdef fmpz_mod_ctx_t _fmpz_mod_one_ctx
-fmpz_one(_fmpz_one)
-fmpz_mod_ctx_init(_fmpz_mod_one_ctx, _fmpz_one)
 
 cdef class fmpz_mod_poly_ctx:
     """
@@ -264,15 +258,18 @@ cdef class fmpz_mod_poly(flint_poly):
     """
     """
     def __cinit__(self):
-        fmpz_mod_poly_init(self.val, _fmpz_mod_one_ctx)
+        self.initialized = False
 
     def __dealloc__(self):
-        fmpz_mod_poly_clear(self.val, self.ctx.mod.val)
+        if self.initialized:
+            fmpz_mod_poly_clear(self.val, self.ctx.mod.val)
 
     def __init__(self, val, ctx):
         if not typecheck(ctx, fmpz_mod_poly_ctx):
             raise TypeError
         self.ctx = ctx
+        fmpz_mod_poly_init(self.val, self.ctx.mod.val)
+        self.initialized = True
         self.ctx.set_any_as_fmpz_mod_poly(self.val, val)
 
     # ---------------- #
```
This increases the base size of an `fmpz_mod_poly` from 72 to 80 bytes. Not sure if there is a better way to arrange the structs to reduce that.

Reply:
Does having initalized as a bint help?

Reply:
Oh I didn't know if defining the initisation as a `bint` using `cdef` would be differentt than a `python` `bool`. 

I didnt think that

```py
cdef bint initialized
```

Would be the same as

```py
self.initialized = False
```

Reply:
At the C level `bint` is just a (C) `int`. Cython uses `bint` as a type to handle automatic conversions from Python's `bool` to C's `int`. This is what the C code for the struct looks like:
```c
 * cdef class fmpz_mod_poly(flint_poly):             # <<<<<<<<<<<<<<
 *     cdef fmpz_mod_poly_t val
 *     cdef fmpz_mod_poly_ctx ctx
 */
struct __pyx_obj_5flint_5types_13fmpz_mod_poly_fmpz_mod_poly {
  struct __pyx_obj_5flint_10flint_base_10flint_base_flint_poly __pyx_base;
  struct __pyx_vtabstruct_5flint_5types_13fmpz_mod_poly_fmpz_mod_poly *__pyx_vtab;
  fmpz_mod_poly_t val;
  struct __pyx_obj_5flint_5types_13fmpz_mod_poly_fmpz_mod_poly_ctx *ctx;
  int initialized;
};
```
So we have:

- `base_flint_poly` struct for the superclass
- v table pointer
- `fmpz_mod_poly_t` inline
- pointer to the Python `fmpz_mod_poly_ctx` object.
- The int `initialized`.

Possibly reordering those things would mean that `bint` could use less space but I think it depends on the sizes of the different structs and alignment.

Reply:
Addressed this in the latest commit, and added some more functions too